File: cloudyview/soar/track.py
"""Flight-track recording and offline video re-rendering.

The in-app recorder (R key) captures only the *track* — per-frame
(t, camera) samples plus a scene header — as a small JSON file. The
offline pass resamples that track at an exact output frame rate
(non-uniform Catmull-Rom through the hand-flown samples, which also
smooths low-fps input) and renders each video frame with full converged
temporal accumulation, so the result has none of the in-flight motion
speckle. Frames stream straight into ffmpeg; no intermediate files.

Track schema: {"schema": "cloudyview.track.v1",
               "header": <render_metadata dict>,
               "samples": [[t, x, y, z, azimuth, elevation, fov], ...]}
with camera fields in the Camera relative-coordinate convention.
"""

from __future__ import annotations

import json
import shutil
import subprocess
from pathlib import Path
from time import perf_counter
from typing import Sequence

import numpy as np

from ..camera import Camera

TRACK_SCHEMA = "cloudyview.track.v1"
# Relative x/y span the domain as [-1, 1]; a periodic flight-path wrap
# shows up as a near-full-span jump between consecutive samples.
_REL_PERIOD = 2.0
_WRAP_JUMP_THRESHOLD = 1.0


def save_track(path: str | Path, header: dict, samples: Sequence[Sequence[float]]) -> Path:
    """Write a recorded track; returns the path."""
    path = Path(path)
    payload = {
        "schema": TRACK_SCHEMA,
        "header": header,
        "samples": [[float(v) for v in s] for s in samples],
    }
    path.write_text(json.dumps(payload))
    return path


def load_track(path: str | Path) -> tuple[dict, np.ndarray]:
    """Read a track file -> (header, samples[n, 7])."""
    payload = json.loads(Path(path).read_text())
    schema = payload.get("schema")
    if schema != TRACK_SCHEMA:
        raise ValueError(
            f"{path}: expected schema {TRACK_SCHEMA!r}, got {schema!r}."
        )
    samples = np.asarray(payload["samples"], dtype=np.float64)
    if samples.ndim != 2 or samples.shape[1] != 7:
        raise ValueError(
            f"{path}: samples must be [n, 7] (t, x, y, z, az, el, fov); "
            f"got shape {samples.shape}."
        )
    if samples.shape[0] < 2:
        raise ValueError(f"{path}: need at least 2 samples to resample.")
    return payload["header"], samples


def _unwrap_periodic(values: np.ndarray, period: float,
                     threshold: float) -> np.ndarray:
    """Make a wrapped coordinate continuous across period jumps."""
    out = values.astype(np.float64).copy()
    jumps = np.diff(out)
    correction = np.zeros_like(out)
    correction[1:] = np.cumsum(
        np.where(jumps > threshold, -period,
                 np.where(jumps < -threshold, period, 0.0))
    )
    return out + correction


def _catmull_rom(times: np.ndarray, values: np.ndarray,
                 t_out: np.ndarray) -> np.ndarray:
    """Non-uniform (time-parameterized) Catmull-Rom through all samples.

    Barry-Goldman formulation with the real sample times as knots, so
    irregular in-flight frame timing interpolates correctly. Endpoints are
    handled by clamping the outer control points.
    """
    n = len(times)
    idx = np.clip(np.searchsorted(times, t_out, side="right") - 1, 0, n - 2)
    out = np.empty_like(t_out)
    for k, (i, t) in enumerate(zip(idx, t_out)):
        i0, i1, i2, i3 = max(i - 1, 0), i, i + 1, min(i + 2, n - 1)
        t0, t1, t2, t3 = times[i0], times[i1], times[i2], times[i3]
        p0, p1, p2, p3 = values[i0], values[i1], values[i2], values[i3]
        # Degenerate knot spacing (clamped ends / duplicate stamps) falls
        # back to linear inside the segment.
        if t2 <= t1:
            out[k] = p1
            continue

        def lerp(pa, pb, ta, tb):
            if tb <= ta:
                return pa
            w = (t - ta) / (tb - ta)
            return pa + (pb - pa) * w

        a1 = lerp(p0, p1, t0, t1)
        a2 = lerp(p1, p2, t1, t2)
        a3 = lerp(p2, p3, t2, t3)
        b1 = lerp(a1, a2, t0, t2)
        b2 = lerp(a2, a3, t1, t3)
        out[k] = lerp(b1, b2, t1, t2)
    return out


def resample_track(samples: np.ndarray, fps: float, *,
                   periodic: bool = True) -> list[tuple[float, Camera]]:
    """Resample hand-flown samples at exact 1/fps steps -> [(t, Camera)].

    Azimuth is unwrapped before interpolation (359 -> 1 goes through 0);
    in periodic domains the x/y flight-path wrap is unwrapped the same way
    and re-wrapped into [-1, 1) afterwards.
    """
    if fps <= 0:
        raise ValueError(f"fps must be positive; got {fps!r}.")
    order = np.argsort(samples[:, 0], kind="stable")
    samples = samples[order]
    keep = np.ones(len(samples), dtype=bool)
    keep[1:] = np.diff(samples[:, 0]) > 0
    samples = samples[keep]
    if len(samples) < 2:
        raise ValueError("track collapses to <2 unique-time samples.")

    times = samples[:, 0]
    x = samples[:, 1]
    y = samples[:, 2]
    if periodic:
        x = _unwrap_periodic(x, _REL_PERIOD, _WRAP_JUMP_THRESHOLD)
        y = _unwrap_periodic(y, _REL_PERIOD, _WRAP_JUMP_THRESHOLD)
    az = np.deg2rad(samples[:, 4])
    az = np.rad2deg(np.unwrap(az))

    t_out = np.arange(times[0], times[-1] + 1e-9, 1.0 / fps)
    cols = {
        "x": _catmull_rom(times, x, t_out),
        "y": _catmull_rom(times, y, t_out),
        "z": _catmull_rom(times, samples[:, 3], t_out),
        "az": _catmull_rom(times, az, t_out),
        "el": np.clip(_catmull_rom(times, samples[:, 5], t_out), -90.0, 90.0),
        "fov": _catmull_rom(times, samples[:, 6], t_out),
    }
    if periodic:
        for key in ("x", "y"):
            cols[key] = (cols[key] + 1.0) % _REL_PERIOD - 1.0

    frames = []
    for k, t in enumerate(t_out):
        frames.append((
            float(t),
            Camera(
                position=(cols["x"][k], cols["y"][k], cols["z"][k]),
                azimuth=float(cols["az"][k] % 360.0),
                elevation=float(cols["el"][k]),
                fov=float(cols["fov"][k]),
            ),
        ))
    return frames


class TrackVideoRender:
    """A track -> mp4 render the caller drives one frame at a time.

    Foreground by design. The windowed app steps this from its draw loop
    rather than handing it to a thread: the renderer and its resident
    volume are not shareable across threads, and giving the GPU wholly to
    the encode is the point — a live progress bar costs one blit a frame.

    Each output frame is rendered offscreen with `accumulate_frames`
    jittered accumulation passes — the converged-still quality the app only
    reaches when the camera stops. Frames are piped to ffmpeg as raw RGB;
    requires the `ffmpeg` binary (no fallback encoder).

    renderer : InteractiveRenderer, optional
        Render with an already-resident volume instead of loading the field
        again. The app passes its own: a second copy of a gigavoxel field
        would not fit beside the first. Its quality tier is set to High for
        the render and restored on close.
    """

    def __init__(self, track_path, out_path, *, fps: float = 60.0,
                 size: tuple[int, int] = (1920, 1080),
                 accumulate_frames: int = 24, crf: int = 18,
                 renderer=None):
        if shutil.which("ffmpeg") is None:
            raise RuntimeError(
                "ffmpeg not found on PATH — install it (e.g. `sudo dnf "
                "install ffmpeg`) to encode track videos. Refusing to fall "
                "back to writing frame files."
            )
        from ..cloudfield import load
        from .engine import InteractiveRenderer

        self.track_path = Path(track_path)
        self.out_path = Path(out_path)
        header, samples = load_track(self.track_path)

        source = header.get("source", {})
        liquid = source.get("path")
        if not liquid:
            raise ValueError(f"{self.track_path}: header has no source path.")
        render_opts = header.get("render", {})
        sun = header.get("sun", {})

        # yuv420p wants even dimensions.
        self.width, self.height = (int(size[0]) & ~1, int(size[1]) & ~1)
        self.fps = float(fps)
        self.accumulate_frames = int(accumulate_frames)

        self._restore_tier = None
        if renderer is None:
            print(f"track: loading {liquid}", flush=True)
            field = load(
                liquid,
                ice=source.get("ice_path"),
                liquid_water_var=source.get("liquid_var"),
                ice_water_var=source.get("ice_var"),
            )
            renderer = InteractiveRenderer(
                field,
                periodic=bool(render_opts.get("periodic", True)),
                extinction_multiplier=float(
                    render_opts.get("extinction_multiplier", 1.0)
                ),
                volume_fp16=bool(render_opts.get("volume_fp16", False)),
            )
        else:
            self._restore_tier = renderer.quality_tier
        renderer.set_quality_tier("high", camera_moving=False)
        self.renderer = renderer

        self._frames = resample_track(
            samples, self.fps, periodic=renderer.periodic
        )
        self.total = len(self._frames)
        self.frames_done = 0

        self._uniform_kwargs = {
            "sun_azimuth": float(sun.get("azimuth", 235.0)),
            "sun_elevation": float(sun.get("elevation", 25.0)),
        }
        for key in (
            "gradient_shading_strength", "deep_shadow_ms_suppression",
            "ambient_occlusion_strength", "bounce_depth_attenuation",
            "light_march_lod_degrees", "view_step_lod_degrees",
        ):
            if key in render_opts:
                self._uniform_kwargs[key] = float(render_opts[key])

        cmd = [
            "ffmpeg", "-y", "-hide_banner", "-loglevel", "error",
            "-f", "rawvideo", "-pix_fmt", "rgb24",
            "-s", f"{self.width}x{self.height}",
            "-r", f"{self.fps}", "-i", "-",
            "-c:v", "libx264", "-crf", str(int(crf)), "-preset", "medium",
            "-pix_fmt", "yuv420p", "-movflags", "+faststart",
            str(self.out_path),
        ]
        self._proc = subprocess.Popen(cmd, stdin=subprocess.PIPE)
        self._t_start = perf_counter()
        self._closed = False

    @property
    def done(self) -> bool:
        return self.frames_done >= self.total

    def step(self, n: int = 1) -> bool:
        """Render up to n frames. Returns True once every frame is encoded."""
        for _ in range(int(n)):
            if self.done:
                break
            _, camera = self._frames[self.frames_done]
            try:
                image = self.renderer.render(
                    camera, size=(self.width, self.height), jitter=True,
                    accumulate_frames=self.accumulate_frames,
                    **self._uniform_kwargs,
                )
                self._proc.stdin.write(
                    np.ascontiguousarray(image[:, :, :3]).tobytes()
                )
            except BaseException:
                self.abort()
                raise
            self.frames_done += 1
        return self.done

    def progress(self) -> dict:
        elapsed = perf_counter() - self._t_start
        rate = self.frames_done / max(elapsed, 1e-6)
        remaining = self.total - self.frames_done
        return {
            "frame": self.frames_done,
            "total": self.total,
            "percent": 100.0 * self.frames_done / max(self.total, 1),
            "fps": rate,
            "eta": remaining / rate if rate > 0 else None,
            "elapsed": elapsed,
        }

    def _restore(self) -> None:
        if self._restore_tier is not None:
            self.renderer.set_quality_tier(
                self._restore_tier, camera_moving=False
            )
            self._restore_tier = None

    def abort(self) -> None:
        """Stop encoding and leave no half-written video behind."""
        if self._closed:
            return
        self._closed = True
        try:
            self._proc.stdin.close()
            self._proc.terminate()
        finally:
            self._restore()
            self.out_path.unlink(missing_ok=True)

    def close(self) -> Path:
        """Finalize the encode; raises if ffmpeg failed."""
        if self._closed:
            return self.out_path
        self._closed = True
        try:
            self._proc.stdin.close()
            code = self._proc.wait()
        finally:
            self._restore()
        if code != 0:
            raise RuntimeError(f"ffmpeg exited with status {code}.")
        duration = self.total / self.fps
        print(
            f"track: wrote {self.out_path} ({self.total} frames, "
            f"{duration:.1f}s at {self.fps:g} fps)",
            flush=True,
        )
        return self.out_path


def render_track(track_path: str | Path, out_path: str | Path, *,
                 fps: float = 60.0, size: tuple[int, int] = (1920, 1080),
                 accumulate_frames: int = 24,
                 crf: int = 18,
                 renderer=None,
                 progress_callback=None) -> Path:
    """Re-render a recorded track into a video (blocks; prints progress).

    The CLI entry point: drives :class:`TrackVideoRender` to completion.
    """
    render = TrackVideoRender(
        track_path, out_path, fps=fps, size=size,
        accumulate_frames=accumulate_frames, crf=crf, renderer=renderer,
    )
    while not render.step():
        progress = render.progress()
        if progress_callback is not None:
            progress_callback(progress)
        if render.frames_done % 30 == 0:
            eta = progress["eta"]
            print(
                f"track: frame {progress['frame']}/{progress['total']} "
                f"({progress['fps']:.1f} fps render, ETA "
                f"{(eta or 0.0) / 60.0:.1f} min)",
                flush=True,
            )
    if progress_callback is not None:
        progress_callback(render.progress())
    return render.close()

File: web/soar/track.js
// Flight-track recording and resampling. Port of cloudyview/soar/track.py.
//
// Recording captures only the TRACK — per-frame (t, camera) samples plus the
// render header — not pixels. The video pass then resamples that track at an
// exact output frame rate and renders each frame with full converged temporal
// accumulation, so the result has none of the in-flight motion speckle. That
// is the whole reason a screen recording is not the same thing: a hand-flown
// sample stream is irregular, and every output frame here takes as long as it
// takes without changing the video's timing by a millisecond.
//
// Track schema is shared with the desktop, so a track recorded in the browser
// can be re-rendered by `render_track` in Python and vice versa:
//   {"schema": "cloudyview.track.v2",
//    "header": <render metadata>,
//    "samples": [[t, x, y, z, azimuth, elevation, fov, sx, sy], ...]}
// with x/y/z in the relative-coordinate convention — the CLOUD frame, folded
// at the cloud period, exactly as v1 wrote them — and sx/sy the camera's
// position in the scene's surface-tile frame, in TILE-RELATIVE units
// (surfacePosition / tileExtent, so the period is 1.0 and the schema needs
// no tile size). The two frames wrap at independent periods, which is why a
// sample carries both: fold one into the other and a recorded city flight
// replays over the wrong district — or the day ocean teleports under a
// replayed crossing. Every scene's camera carries a surface frame (the day
// ocean's included), so v2 samples are 9 columns with the first seven
// byte-identical to v1's; a camera without one writes 7 columns, and a v1
// track loads and resamples the same way.

"use strict";

export const TRACK_SCHEMA = "cloudyview.track.v2";

// Relative x/y span the domain as [-1, 1], so a periodic flight-path wrap is
// a near-full-span jump between consecutive samples and nothing else is.
const REL_PERIOD = 2.0;
const WRAP_JUMP_THRESHOLD = 1.0;

/** Make a wrapped coordinate continuous across period jumps. */
export function unwrapPeriodic(values, period = REL_PERIOD,
                               threshold = WRAP_JUMP_THRESHOLD) {
  const out = Float64Array.from(values);
  let correction = 0.0;
  for (let i = 1; i < out.length; i++) {
    const jump = values[i] - values[i - 1];
    if (jump > threshold) correction -= period;
    else if (jump < -threshold) correction += period;
    out[i] += correction;
  }
  return out;
}

/**
 * Angles in degrees, made continuous — numpy's unwrap, in degrees.
 *
 * Without this, flying through north makes azimuth jump 359 -> 1 and the
 * interpolated camera spins all the way round the other way in one frame.
 */
export function unwrapDegrees(values) {
  const out = Float64Array.from(values);
  let correction = 0.0;
  for (let i = 1; i < values.length; i++) {
    const d = values[i] - values[i - 1];
    let dmod = (((d + 180.0) % 360.0) + 360.0) % 360.0 - 180.0;
    if (dmod === -180.0 && d > 0.0) dmod = 180.0;
    if (Math.abs(d) >= 180.0) correction += dmod - d;
    out[i] = values[i] + correction;
  }
  return out;
}

/**
 * Non-uniform (time-parameterized) Catmull-Rom through every sample.
 *
 * Barry-Goldman, with the real sample times as knots, so irregular in-flight
 * frame timing interpolates correctly instead of being treated as evenly
 * spaced — which is what turns a stutter in the recording into a lurch in the
 * video. Endpoints clamp their outer control points.
 */
export function catmullRom(times, values, tOut) {
  const n = times.length;
  const out = new Float64Array(tOut.length);
  let i = 0;                     // tOut is ascending, so the knot only marches
  for (let k = 0; k < tOut.length; k++) {
    const t = tOut[k];
    while (i + 1 < n && times[i + 1] <= t) i++;
    i = Math.min(Math.max(i, 0), n - 2);

    const i0 = Math.max(i - 1, 0), i1 = i, i2 = i + 1, i3 = Math.min(i + 2, n - 1);
    const t0 = times[i0], t1 = times[i1], t2 = times[i2], t3 = times[i3];
    const p0 = values[i0], p1 = values[i1], p2 = values[i2], p3 = values[i3];
    // Degenerate knot spacing (clamped ends, duplicate stamps) falls back to
    // linear inside the segment.
    if (t2 <= t1) { out[k] = p1; continue; }

    const lerp = (pa, pb, ta, tb) =>
      tb <= ta ? pa : pa + (pb - pa) * ((t - ta) / (tb - ta));

    const a1 = lerp(p0, p1, t0, t1);
    const a2 = lerp(p1, p2, t1, t2);
    const a3 = lerp(p2, p3, t2, t3);
    const b1 = lerp(a1, a2, t0, t2);
    const b2 = lerp(a2, a3, t1, t3);
    out[k] = lerp(b1, b2, t1, t2);
  }
  return out;
}

/**
 * Resample hand-flown samples at exact 1/fps steps.
 *
 * Returns `[{t, position, azimuth, elevation, fov, surfacePosition}, ...]`
 * with position in relative coordinates and surfacePosition tile-relative
 * (or null — see below). Azimuth is unwrapped before interpolation so 359 to
 * 1 goes through 0 rather than the long way round; in a periodic domain each
 * wrapped position column is unwrapped the same way at its own frame's
 * period and re-wrapped afterwards.
 */
const frameCountBetween = (t0, t1, fps) =>
  Math.floor((t1 - t0 + 1e-9) * fps) + 1;

/**
 * How many frames `resampleTrack` would produce — without producing them.
 *
 * The menu shows this number the instant a recording stops, purely to say
 * "which becomes 900 frames at 30 fps". Resampling to count was fine while a
 * track was a minute of flying; it is not fine for a track whose length is
 * whatever the clock did, and it is the arithmetic below either way.
 */
export function resampledFrameCount(samples, fps) {
  const times = distinctAscendingTimes(samples, fps);
  return frameCountBetween(times[0], times[times.length - 1], fps);
}

/** The sample times a resample would actually interpolate over. */
function distinctAscendingTimes(samples, fps) {
  if (!(fps > 0)) throw new Error(`fps must be positive; got ${fps}.`);
  const sorted = samples.map((s) => s[0]).sort((a, b) => a - b);
  const distinct = sorted.filter((t, i) => i === 0 || t > sorted[i - 1]);
  if (distinct.length < 2) {
    throw new Error(
      `The track collapses to ${distinct.length} sample(s) with distinct ` +
      "times, which is not enough to interpolate. Fly for longer.");
  }
  return distinct;
}

export function resampleTrack(samples, fps, { periodic = true } = {}) {
  if (!(fps > 0)) throw new Error(`fps must be positive; got ${fps}.`);

  const sorted = samples.map((s) => Array.from(s))
    .sort((a, b) => a[0] - b[0]);
  const unique = sorted.filter((s, i) => i === 0 || s[0] > sorted[i - 1][0]);
  if (unique.length < 2) {
    throw new Error(
      `The track collapses to ${unique.length} sample(s) with distinct ` +
      "times, which is not enough to interpolate. Fly for longer.");
  }
  // 9 columns is a v2 sample (every scene's camera carries a surface frame
  // now, the day ocean's included); 7 is a v1 recording — those predate the
  // surface frame and replay with zero drift (the old behavior, not
  // repairable: the fold already lost which tile copy the flight was in).
  // A mix is a corrupt track, not a choice to make quietly.
  const widths = new Set(unique.map((s) => s.length));
  if (widths.size !== 1 || !(widths.has(7) || widths.has(9))) {
    throw new Error(
      "Track samples must uniformly have 7 columns (v1 / no surface frame) " +
      `or 9 (v2 with one); got widths {${[...widths].join(", ")}}.`);
  }
  const hasSurface = widths.has(9);

  const column = (j) => unique.map((s) => s[j]);
  const times = column(0);
  // Each column unwraps and folds at ITS OWN frame's period: the cloud
  // frame's x/y at REL_PERIOD as ever, the surface frame's sx/sy at 1.0
  // (tile-relative units). Callers do not choose periods — the columns carry
  // their frames, and folding either at the other's period is precisely the
  // district drift the two-frame schema exists to prevent.
  let x = column(1), y = column(2);
  if (periodic) { x = unwrapPeriodic(x); y = unwrapPeriodic(y); }
  const az = unwrapDegrees(column(4));
  let sx = null, sy = null;
  if (hasSurface && periodic) {
    sx = unwrapPeriodic(column(7), 1.0, 0.5);
    sy = unwrapPeriodic(column(8), 1.0, 0.5);
  } else if (hasSurface) {
    sx = column(7); sy = column(8);
  }

  const count = frameCountBetween(times[0], times[times.length - 1], fps);
  const t0 = times[0];
  const tOut = new Float64Array(count);
  for (let k = 0; k < count; k++) tOut[k] = t0 + k / fps;

  const cols = {
    x: catmullRom(times, x, tOut),
    y: catmullRom(times, y, tOut),
    z: catmullRom(times, column(3), tOut),
    az: catmullRom(times, az, tOut),
    el: catmullRom(times, column(5), tOut),
    fov: catmullRom(times, column(6), tOut),
    ...(hasSurface ? { sx: catmullRom(times, sx, tOut),
                       sy: catmullRom(times, sy, tOut) } : {}),
  };
  if (periodic) {
    // Fold back at the same period each unwrap used: x/y centred into
    // [-1, 1) bit for bit as v1 did, sx/sy into [0, 1) — the tile frame's
    // own convention.
    for (const key of ["x", "y"]) {
      for (let k = 0; k < count; k++) {
        cols[key][k] = (((cols[key][k] + 1.0) % REL_PERIOD) + REL_PERIOD)
                       % REL_PERIOD - 1.0;
      }
    }
    for (const key of hasSurface ? ["sx", "sy"] : []) {
      for (let k = 0; k < count; k++) {
        cols[key][k] = ((cols[key][k] % 1.0) + 1.0) % 1.0;
      }
    }
  }

  const frames = [];
  for (let k = 0; k < count; k++) {
    frames.push({
      t: tOut[k],
      position: [cols.x[k], cols.y[k], cols.z[k]],
      azimuth: ((cols.az[k] % 360.0) + 360.0) % 360.0,
      elevation: Math.min(90.0, Math.max(-90.0, cols.el[k])),
      fov: cols.fov[k],
      // Tile-relative, like the samples; null on a track with no surface
      // frame, and the replay then folds against the static offset.
      surfacePosition: hasSurface ? [cols.sx[k], cols.sy[k]] : null,
    });
  }
  return frames;
}

/**
 * A track long enough that it is no longer a flight anyone meant to record.
 *
 * Ten minutes is 18,000 frames at 30 fps, which is already a big video and a
 * big render. The point of the cap is not the video, though — it is that the
 * clock below has to stop somewhere, and a bound stated in seconds of flying
 * is one a person can reason about.
 */
export const MAX_TRACK_SECONDS = 600.0;

/**
 * The in-flight recorder: appends one sample per rendered frame.
 *
 * Samples are the camera's own relative coordinates, taken after the frame
 * that showed them, so the track describes what was on screen rather than
 * what was about to be.
 *
 * The clock is the flight's, not the wall's: it advances by the delta of each
 * rendered frame. Wall time made "start recording, switch to another app for
 * the afternoon, come back" into a track claiming to be five hours long —
 * requestAnimationFrame does not fire in a hidden tab, so no samples describe
 * those hours, but the first one afterwards carried the whole gap in its
 * timestamp. Resampling that at 30 fps asks for half a million frames.
 */
export class TrackRecorder {
  constructor() {
    this.samples = [];
    this.elapsed = 0.0;
    this._recording = false;
  }

  get recording() { return this._recording; }
  get duration() {
    return this.samples.length ? this.samples[this.samples.length - 1][0] : 0.0;
  }
  /** True once the flight has run past what a track is allowed to hold. */
  get full() { return this.elapsed >= MAX_TRACK_SECONDS; }

  start() {
    this.samples = [];
    this.elapsed = 0.0;
    this._recording = true;
  }

  /** Advance the flight clock by one rendered frame. */
  advance(deltaSeconds) {
    if (!this._recording) return;
    this.elapsed += Math.max(0.0, deltaSeconds);
  }

  /** One sample at the current flight time. `camera` is the live FlightCamera. */
  sample(camera) {
    if (!this._recording) return;
    const rel = camera.relativePosition();
    const s = [
      this.elapsed,
      rel[0], rel[1], rel[2],
      camera.azimuth, camera.elevation, camera.fov,
    ];
    // The surface frame, tile-relative: x/y above are folded at the cloud
    // period, which loses which tile copy the flight is in — sx/sy is where
    // that lives (the city district, or the day ocean's phase). Appended, so
    // the first seven columns stay byte-identical to a v1 sample.
    if (camera.surfacePosition) {
      s.push(camera.surfacePosition[0] / camera.surfaceTileExtent,
             camera.surfacePosition[1] / camera.surfaceTileExtent);
    }
    this.samples.push(s);
  }

  stop() {
    this._recording = false;
    return this.samples;
  }
}

/** The JSON a track file holds, ready for download or for Python. */
export function trackPayload(header, samples) {
  return {
    schema: TRACK_SCHEMA,
    header,
    samples: samples.map((s) => Array.from(s, Number)),
  };
}
